lib/src/test/java/org/kpipe/consumer/OffsetManagerTest.java

```diff
@@ -0,0 +1,798 @@
+          } finally {
+            completionLatch.countDown();
+          }
+        })
+          .start();
+      }
+
+      // Start all threads simultaneously
+      startLatch.countDown();
+
+      // Wait for all threads to complete
+      assertTrue(completionLatch.await(3, TimeUnit.SECONDS), "All threads should complete in time");
+
+      // Give time for offsets to be processed
+      Thread.sleep(200);
+
+      // Force final commit and verify
+      concurrentManager.commitSyncAndWait(1);
+    }
+  }
+
+  @Nested
+  class ConfigurationTests {
+
+    @Test
+    void shouldRespectBuilderConfigurations() {
+      // Create a manager with custom configuration
+      final var customManager = OffsetManager
+        .builder(mockConsumer)
+        .withCommandQueue(commandQueue)
+        .withCommitInterval(Duration.ofMillis(200))
+        .build();
+
+      // Just verify it can be created and closed without exceptions
+      assertDoesNotThrow(() -> {
+        customManager.trackOffset(partition, 101L);
+        customManager.markOffsetProcessed(partition, 101L);
+        customManager.close();
+      });
+    }
+  }
+
+  private ConsumerCommand performCommitAndCaptureCommand() throws Exception {
+    final var commandLatch = new CountDownLatch(1);
+    final var commandRef = new AtomicReference<ConsumerCommand>();
+
+    // Start a command listener thread
+    Thread
+      .ofPlatform()
+      .daemon()
+      .start(() -> {
+        try {
+          final var cmd = commandQueue.poll(30, TimeUnit.SECONDS);
+          if (cmd != null) {
+            commandRef.set(cmd);
+            commandLatch.countDown();
+          }
+        } catch (final InterruptedException e) {
+          // Ignore interruption
+        }
+      });
+
+    // Start the commit operation
+    CompletableFuture.runAsync(() -> {
+      try {
+        offsetManager.commitSyncAndWait(30);
+      } catch (final InterruptedException e) {
+        // Ignore interruption
+      }
+    });
+
+    // Wait for command
+    final var commandReceived = commandLatch.await(10, TimeUnit.SECONDS);
+    assertTrue(commandReceived, "Command should be produced within timeout");
+
+    final var command = commandRef.get();
+    assertNotNull(command, "Commit command should be generated");
+
+    return command;
+  }
+
+  private void assertEventually(final Supplier<Boolean> condition, final Duration timeout, final String message)
+    throws Exception {
+    final var deadline = System.currentTimeMillis() + timeout.toMillis();
+
+    while (System.currentTimeMillis() < deadline) {
+      if (condition.get()) {
+        return;
+      }
+      Thread.sleep(50);
+    }
+
+    fail(message);
+  }
+
+  static Stream<Arguments> contiguousOffsetScenarios() {
+    return Stream.of(
+      Arguments.of(
+        "Basic gap scenario - process offsets out of order with a missing offset",
+        100L, // initialOffset
+        List.of(101L, 105L, 102L, 103L, 104L), // offsetsToTrack
```

Comment:
technically this should be impossible - offsets are submitted by single thread now - so that would be in incremental order only...

Reply:
removed